ethcore/src/client/client.rs

```diff
@@ -1652,6 +1652,32 @@ impl MiningBlockChainClient for Client {
 		open_block
 	}
 
+	fn reopen_block(&self, block: ClosedBlock) -> OpenBlock {
```

Comment:
This is `MiningBlockChainClient` so it does indeed contain some mining-specific logic. See `prepare_open_block` above for example.

As for fetching all valid uncles - a chain ancestry iteration is required to check if an uncle is valid. It is actually faster to iterate once and check if uncle is already in the block.
I've added a small change to avoid reading headers unnecessarily though.